### PR TITLE
AzureCP PR - Fix Extension ID Issue #387

### PR DESCRIPTION
**Pull request by huangpf**
AzureCP PR - Fix Extension ID Issue #387

### CODE REVIEW

**Comment by huangpf** [comment]
@hyonholee @hovsepm Please help review.

File: src/ServiceManagement/Compute/Commands.ServiceManagement/Extensions/Common/ExtensionRole.cs [hunk 1]
```diff
@@ -22,12 +23,28 @@ public class ExtensionRole
...
+            var match = disallowedCharactersRegex.Match(roleName);
+
+            while (match.Success)
+            {
+                roleName = roleName.Remove(match.Index, match.Length);
+                match = disallowedCharactersRegex.Match(roleName);
+            }
+
+            return roleName;
+        }
```
Comment (hovsepm) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/Extensions/Common/ExtensionRole.cs]: this method seems too complex for the thing it is doing. 
Just put following on line 37 and you should be good:
```
return disallowedCharactersRegex.Replace(roleName,"");
```

File: src/ServiceManagement/Compute/Commands.ServiceManagement/Extensions/Common/ExtensionRole.cs [hunk 2]
```diff
@@ -63,12 +78,16 @@ public override string ToString()
...
-            normalizedExtName = normalizedExtName.Replace("_", string.Empty);
+            var normalizedExtName = RemoveDisallowedCharacters(extensionName);
 
             var suffix = new StringBuilder();
             suffix.AppendFormat(ExtensionIdSuffixTemplate, normalizedExtName, slot, index);
+            if (suffix.Length > MaxSuffixLength)
+            {
+                int lenDiff = suffix.Length - MaxSuffixLength;
+                int startIndex = 1; // Suffix starts with '-'
+                suffix.Remove(startIndex + normalizedExtName.Length - lenDiff, lenDiff);
```
Comment (hovsepm) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/Extensions/Common/ExtensionRole.cs]: what do you want to achieve here? string.Remove returns a new string which you do not use anywhere.

Comment (hovsepm) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/Extensions/Common/ExtensionRole.cs]: BTW, please cover this case in the tests as well ;)

File: src/ServiceManagement/Compute/Commands.ServiceManagement/Extensions/Common/ExtensionRole.cs [hunk 3]
```diff
@@ -63,12 +78,16 @@ public override string ToString()
...
+            }
```
Comment (hovsepm) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/Extensions/Common/ExtensionRole.cs]: is this for limiting the string length? If so, this should be enough 
```
suffix = siffix.Substring(0, MaxSuffixLength);
```
right?

Comment (huangpf) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/Extensions/Common/ExtensionRole.cs]: Actually no... So the format of the extension ID is ```{role-name-part} + {suffix}```, where ```{suffix} = -{extension-name-part}-{slot}-Ext-{num}``` (all from legacy design request).

I want to limit the ```suffix``` length, but I can't simply cut it off by ```MaxSuffixLength```, and I have to truncate the ```{extension-name-part}```.

File: src/ServiceManagement/Compute/Commands.ServiceManagement/Extensions/Common/ExtensionRole.cs (hunk 2 above)
Comment (huangpf) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/Extensions/Common/ExtensionRole.cs]: Yes, it's actually covered, :-)... -- The test case is to validate that a very long extension name will be truncated in the suffix, while the other parts, i.e. slot & number are preserved.

**Comment by huangpf** [comment]
Ping... @hovsepm

**Comment by markcowl** [comment]
@huangpf please fix merge conflicts, pull the latest, and re-generate the wix file.

**Comment by huangpf** [comment]
@markcowl Done~

**Pull request merged**